ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+}
+
+func (client *ecsClient) setInstanceIdentity(
+	registerRequest ecsmodel.RegisterContainerInstanceInput) ecsmodel.RegisterContainerInstanceInput {
+	instanceIdentityDoc := ""
+	instanceIdentitySignature := ""
+
+	if client.configAccessor.NoInstanceIdentityDocument() {
+		logger.Info("Fetching Instance ID Document has been disabled")
+		registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
+		registerRequest.InstanceIdentityDocumentSignature = &instanceIdentitySignature
+		return registerRequest
+	}
+
+	iidRetrieved := true
+	backoff := retry.NewExponentialBackoff(setInstanceIdRetryBackoffMin, setInstanceIdRetryBackoffMax,
+		setInstanceIdRetryBackoffJitter, setInstanceIdRetryBackoffMultiple)
+	ctx, cancel := context.WithTimeout(context.Background(), setInstanceIdRetryTimeOut)
+	defer cancel()
+	err := retry.RetryWithBackoffCtx(ctx, backoff, func() error {
+		var attemptErr error
+		logger.Debug("Attempting to get Instance Identity Document")
+		instanceIdentityDoc, attemptErr = client.ec2metadata.GetDynamicData(ec2.InstanceIdentityDocumentResource)
+		if attemptErr != nil {
+			logger.Debug("Unable to get instance identity document, retrying", logger.Fields{
+				field.Error: attemptErr,
+			})
+			// Force credentials to expire in case they are stale but not expired.
+			client.credentialsProvider.Expire()
+			client.credentialsProvider = instancecreds.GetCredentials(client.configAccessor.External())
+			return apierrors.NewRetriableError(apierrors.NewRetriable(true), attemptErr)
+		}
+		logger.Debug("Successfully retrieved Instance Identity Document")
+		return nil
+	})
+	if err != nil {
+		logger.Error("Unable to get instance identity document", logger.Fields{
+			field.Error: err,
+		})
+		iidRetrieved = false
+	}
+	registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
+
+	if iidRetrieved {
+		instanceIdentitySignature, err = client.ec2metadata.
+			GetDynamicData(ec2.InstanceIdentityDocumentSignatureResource)
+		if err != nil {
+			logger.Error("Unable to get instance identity signature", logger.Fields{
+				field.Error: err,
+			})
+		}
+	}
+
+	registerRequest.InstanceIdentityDocumentSignature = &instanceIdentitySignature
+	return registerRequest
+}
+
+func attributesToMap(attributes []*ecsmodel.Attribute) map[string]string {
+	attributeMap := make(map[string]string)
+	attribs := attributes
+	for _, attribute := range attribs {
+		attributeMap[aws.StringValue(attribute.Name)] = aws.StringValue(attribute.Value)
+	}
+	return attributeMap
+}
+
+func findMissingAttributes(expectedAttributes, actualAttributes map[string]string) ([]string, error) {
+	missingAttributes := make([]string, 0)
+	var err error
+	for key, val := range expectedAttributes {
+		if actualAttributes[key] != val {
+			missingAttributes = append(missingAttributes, key)
+		} else {
+			logger.Trace("Response contained expected value for attribute", logger.Fields{
+				"key": key,
```

Comment:
nit: prefer capital fields?

Reply:
Synced offline, in ecs-agent module thus far we have preferred uncapitalized fields.